### PR TITLE
[dtensor] early return for _split_tensor

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125810

as titled, if _split_tensor does not require padding or even is evenly
sharded on the dim, no need to calculate padding and could simply return

This is to avoid some unnecessary CPU operations

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

**Review by wz337 — APPROVED** [review]
This is great! LGTM!

**Review by XilunWu — APPROVED** [review]

File: torch/distributed/_tensor/placement_types.py [hunk 1]
```diff
@@ -60,9 +60,17 @@ def _split_tensor(
...
+        # chunk tensor over dimension `dim` into n slices
         tensor_list = list(torch.chunk(tensor, num_chunks, dim=self.dim))
-        # compute the chunk size inline with ``torch.chunk``
+
+        # if no need to have padding or tensor dim size is evenly sharded already
+        # we can return early.
+        if not with_padding or tensor.size(self.dim) % num_chunks == 0:
+            if contiguous:
+                tensor_list = [t.contiguous() for t in tensor_list]
+            return tensor_list, []
```
Comment (XilunWu) [inline torch/distributed/_tensor/placement_types.py:71]: directly returning seems problematic according to the test:
```
l = torch.chunk(torch.rand(5, 2), 4)  # yields 3 smaller tensors of size (2, 2), (2, 2), (1, 2)
l[3]  # index out of range
```

**Review by XilunWu — DISMISSED** [review]
Need fix test.

File: torch/distributed/_tensor/placement_types.py (hunk 1 above)
Comment (wz337) [reply torch/distributed/_tensor/placement_types.py:71]: In this case, with_padding would be True and tensor.size(self.dim) % num_chunks won't be 0, right? So it shouldn't go in this if, if I am understanding this correctly.

Comment (XilunWu) [reply torch/distributed/_tensor/placement_types.py:71]: The failed test case has `with_padding=False`.

Comment (wanchaol) [reply torch/distributed/_tensor/placement_types.py]: fixed the tests

**Pull request closed without merging**